ml-proto/host/lexer.mll

```diff
@@ -145,98 +154,181 @@ rule token = parse
+  | (ixx as t)".or"
+    { BINARY (fun (e1, e2) -> intop t (I32_or (e1, e2)) (I64_or (e1, e2))) }
+  | (ixx as t)".xor"
+    { BINARY (fun (e1, e2) -> intop t (I32_xor (e1, e2)) (I64_xor (e1, e2))) }
+  | (ixx as t)".shl"
+    { BINARY (fun (e1, e2) -> intop t (I32_shl (e1, e2)) (I64_shl (e1, e2))) }
+  | (ixx as t)".shr_s"
+    { BINARY (fun (e1, e2) ->
+        intop t (I32_shr_s (e1, e2)) (I64_shr_s (e1, e2))) }
+  | (ixx as t)".shr_u"
+    { BINARY (fun (e1, e2) ->
+        intop t (I32_shr_u (e1, e2)) (I64_shr_u (e1, e2))) }
+  | (fxx as t)".add"
+    { BINARY (fun (e1, e2) -> floatop t (F32_add (e1, e2)) (F64_add (e1, e2))) }
+  | (fxx as t)".sub"
+    { BINARY (fun (e1, e2) -> floatop t (F32_sub (e1, e2)) (F64_sub (e1, e2))) }
+  | (fxx as t)".mul"
+    { BINARY (fun (e1, e2) -> floatop t (F32_mul (e1, e2)) (F64_mul (e1, e2))) }
+  | (fxx as t)".div"
+    { BINARY (fun (e1, e2) -> floatop t (F32_div (e1, e2)) (F64_div (e1, e2))) }
+  | (fxx as t)".min"
+    { BINARY (fun (e1, e2) -> floatop t (F32_min (e1, e2)) (F64_min (e1, e2))) }
+  | (fxx as t)".max"
+    { BINARY (fun (e1, e2) -> floatop t (F32_max (e1, e2)) (F64_max (e1, e2))) }
   | (fxx as t)".copysign"
-    { BINARY (floatop t Float32Op.CopySign Float64Op.CopySign) }
-
-  | (ixx as t)".eq" { COMPARE (intop t Int32Op.Eq Int64Op.Eq) }
-  | (ixx as t)".ne" { COMPARE (intop t Int32Op.Ne Int64Op.Ne) }
-  | (ixx as t)".lt_s" { COMPARE (intop t Int32Op.LtS Int64Op.LtS) }
-  | (ixx as t)".lt_u" { COMPARE (intop t Int32Op.LtU Int64Op.LtU) }
-  | (ixx as t)".le_s" { COMPARE (intop t Int32Op.LeS Int64Op.LeS) }
-  | (ixx as t)".le_u" { COMPARE (intop t Int32Op.LeU Int64Op.LeU) }
-  | (ixx as t)".gt_s" { COMPARE (intop t Int32Op.GtS Int64Op.GtS) }
-  | (ixx as t)".gt_u" { COMPARE (intop t Int32Op.GtU Int64Op.GtU) }
-  | (ixx as t)".ge_s" { COMPARE (intop t Int32Op.GeS Int64Op.GeS) }
-  | (ixx as t)".ge_u" { COMPARE (intop t Int32Op.GeU Int64Op.GeU) }
-  | (fxx as t)".eq" { COMPARE (floatop t Float32Op.Eq Float64Op.Eq) }
-  | (fxx as t)".ne" { COMPARE (floatop t Float32Op.Ne Float64Op.Ne) }
-  | (fxx as t)".lt" { COMPARE (floatop t Float32Op.Lt Float64Op.Lt) }
-  | (fxx as t)".le" { COMPARE (floatop t Float32Op.Le Float64Op.Le) }
-  | (fxx as t)".gt" { COMPARE (floatop t Float32Op.Gt Float64Op.Gt) }
-  | (fxx as t)".ge" { COMPARE (floatop t Float32Op.Ge Float64Op.Ge) }
-
-  | "i64.extend_s/i32" { CONVERT (Values.Int64 Int64Op.ExtendSInt32) }
-  | "i64.extend_u/i32" { CONVERT (Values.Int64 Int64Op.ExtendUInt32) }
-  | "i32.wrap/i64" { CONVERT (Values.Int32 Int32Op.WrapInt64) }
+    { BINARY (fun (e1, e2) ->
+        floatop t (F32_copysign (e1, e2)) (F64_copysign (e1, e2))) }
+
+  | (nxx as t)".select"
+    { SELECT (fun (e1, e2, e3) ->
+        numop t (I32_select (e1, e2, e3)) (I64_select (e1, e2, e3))
+                (F32_select (e1, e2, e3)) (F64_select (e1, e2, e3))) }
+
+  | (ixx as t)".eq"
+    { COMPARE (fun (e1, e2) -> intop t (I32_eq (e1, e2)) (I64_eq (e1, e2))) }
```

Comment:
I'd love to make all this niftier, but I don't know how. It would be nicer if we had curried constructors in OCaml (then it'd be just `COMPARE (intop t I32_eq I64_eq)`), but we don't.